testsuite/tests/parsetree/source.ml

```diff
@@ -156,6 +156,13 @@ module type S = S -> S -> S
 module type S = (S -> S) -> S
 module type S = functor (M : S) -> S -> S
 module type S = (functor (M : S) -> S) -> S
+module type S = (S -> S)[@foo] -> S
+module type S = (functor[@foo] (M : S) -> S) -> S
+
+module type S = sig
+  module rec A : (S with type t = t)
```

Comment:
The parentheses here are a consequence of using `type1` above, but why are they desirable/useful?